cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java

```diff
@@ -361,6 +374,31 @@ private void doMain(String[] args) throws Exception {
     }
   }
 
+  /**
+   * checks for appfabric service path on zookeeper, if they exist, CDAP master is still running, so throw
+   * exception message with information on where its running.
+   * If none of the master service (in HA setup) is running, this method wouldn't throw any exception
+   * @throws Exception
+   */
+  private void ensureCDAPMasterStopped() throws Exception {
+    String appFabricPath = String.format("/discoverable/%s", Constants.Service.APP_FABRIC_HTTP);
+    NodeChildren nodeChildren = zkClientService.getChildren(appFabricPath).get();
+    List<String> runningNodes = new ArrayList<>();
+    // if no children nodes at appfabric path, all master nodes are stopped
+    if (!nodeChildren.getChildren().isEmpty()) {
+      for (String runId : nodeChildren.getChildren()) {
+        // only one children would be present, as only the active master will be registered at this path
+        NodeData nodeData = zkClientService.getData(String.format("%s/%s", appFabricPath, runId)).get();
+        Discoverable discoverable = GSON.fromJson(Bytes.toString(nodeData.getData()), Discoverable.class);
+        runningNodes.add(discoverable.getSocketAddress().getHostName());
+      }
+      String exceptionMessage =
+        String.format("CDAP Master is still running on %s, please stop it before running upgrade.",
+                      com.google.common.base.Joiner.on(",").join(runningNodes));
+      throw new Exception(exceptionMessage);
+    }
+  }
```

Comment:
As a future improvement, the upgrade tool register as a master to become the leader and prevent other masters from starting.